src/txn/txn.c

```diff
@@ -1335,8 +1336,13 @@ __txn_resolve_prepared_op(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit,
              */
             if (head_upd->txnid == WT_TXN_ABORTED)
                 continue;
-            /* Exit once we have visited all updates from the current transaction. */
-            if (head_upd->txnid != txn->id)
+            /*
+             * Exit once we have visited all updates from the current transaction. When a
+             * transaction is claim prepared, we don't assign a txn id to it so the txn id can be 0
+             * which is the same with head_upd if it's restored from disk, so we break if we find a
+             * complete update restored from DS
+             */
+            if (head_upd->txnid != txn->id || F_ISSET(head_upd, WT_UPDATE_RESTORED_FROM_DS))
```

Comment:
Should we also do head_upd->txnid != WT_TXN_NONE && head_upd->txnid != txn->id

Reply:
if `head_upd` is an update restored from disk, we clear the txn id so `head_upd->txnid` would be 0, which is the same as session's txn id, so we wouldn't break if we only break when `head_upd->txnid != WT_TXN_NONE`. I can add the `WT_UPDATE_RESTORED_FROM_HS` flag, or we can break when `if (head_upd->txnid != txn->id || head_upd->prepare_state == WT_PREPARE_INIT` which is cleaner?